### ISSUE TITLE
Compiling infinite goto loop with `-O2` takes clang into an infinite loop

### ISSUE THREAD

**Issue by david-cortes** [llvm/llvm-project#47429]
|  |  |
| --- | --- |
| Bugzilla Link | [48085](https://llvm.org/bz48085) |
| Version | 9.0 |
| OS | Linux |
| Attachments | [Source code that produces the bug](https://user-images.githubusercontent.com/92601585/143761384-46ff9168-ce3a-4a8b-8752-23ea63507310.gz), [example infinite goto loop](https://user-images.githubusercontent.com/92601585/143761385-2421023a-57a9-42b9-bf5c-f0548b4f43f3.gz) |

## Extended Description 

I’ve come across an issue in which, when some C code contains an infinite loop in the form of a ‘goto’ with the exit condition blocked by another `goto`, compiling it with clang will take clang into an infinite loop during compilation, making it get stuck until eventually running out of memory.

Attached is a screenshot of the source code that gave the problem (the original file is this: https://github.com/david-cortes/cmfrec/blob/master/src/collective.c).

I’m additionally attaching a very small reproducible example in which the issue is reproducible, but only with `-O2` and higher.

Steps to reproduce:
- Download the attached source code file
- Compile it with `-O2` - e.g.:
```
    clang -O2 infinite_loop.c
```
Expected behavior: Should compile the source code into an executable, regardless of whether it has an infinite loop or not. Ideally should warn about the infinite loop.

Actual behavior: Compiler gets stuck into an infinite loop trying to optimize that code, until eventually running out of memory.

**Comment by wheatman** [comment]
This was fixed between 11.0.1 and 12.0.0

https://godbolt.org/z/13Y3T59rv

compile with `-O2`

code
```c
#include <stdlib.h>
#include <limits.h>
#include <stdio.h>
#include <time.h>

int main()
{
	int retval = 0;
	double *restrict some_arr = (double*)malloc(100*sizeof(double));
	double *restrict another_arr = (double*)malloc(100*sizeof(double));
	if (some_arr == NULL || another_arr == NULL)
		goto throw_oom;

	srand(time(NULL));
	for (int ix = 0; ix < 100; ix++)
		some_arr[ix] = (double)rand() / (double)RAND_MAX;
	for (int ix = 0; ix < 100; ix++)
		another_arr[ix] = some_arr[ix] * ((double)rand() / (double)RAND_MAX);
	double tot = 0;
	for (int ix = 0; ix < 100; ix++)
		tot += another_arr[ix] * ((double)rand() / (double)RAND_MAX);
	printf("tot is: %.2f\n", tot);
	if (tot < 20)
		goto cleanup;

	cleanup:
		free(some_arr);
		free(another_arr);
	throw_oom:
	{
		retval = 1;
		goto cleanup;
	}
	return retval;
}

```
